File: converted/utilities/p-limit/elide-p-limit.rb
#!/usr/bin/env ruby

# Ruby Integration Example for elide-p-limit
#
# This demonstrates calling the TypeScript p-limit implementation from Ruby
# using Elide's polyglot capabilities.
#
# Benefits:
# - One p-limit implementation shared across TypeScript and Ruby
# - Consistent concurrency limiting behavior across microservices
# - No Ruby library discrepancies
# - Perfect for Rails, Sidekiq, background workers

puts "=== Ruby Calling TypeScript PLimit ===\n"

# NOTE: Exact syntax depends on Elide's Ruby polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Ruby support is ready

# Assuming Elide provides something like:
# p-limit_module = Elide.require('./elide-p-limit.ts')

# Example 1: Basic Usage
# result = p-limit_module.default(input_data)
# puts "Result: #{result}"
# puts ""

# Example 2: Rails Integration
# class DataService
#   def initialize
#     @p-limit = p-limit_module
#   end
#
#   def process(data)
#     @p-limit.default(data)
#   end
# end
#
# service = DataService.new
# result = service.process(input_data)
# puts "Processed: #{result}"
# puts ""

# Example 3: Sidekiq Worker
# class ProcessingJob
#   include Sidekiq::Worker
#
#   def perform(data)
#     result = p-limit_module.default(data)
#     puts "Job result: #{result}"
#   end
# end

puts "Real-world use case:"
puts "- Ruby worker uses p-limit"
puts "- Uses same TypeScript implementation as Node.js API"
puts "- Guarantees consistent concurrency limiting across services"
puts "- Rate limit API calls"
puts ""

puts "Example: Microservices Architecture"
puts "┌─────────────────────────────────────────┐"
puts "│   Elide PLimit (TypeScript)           │"
puts "│   elide-p-limit.ts                        │"
puts "└─────────────────────────────────────────┘"
puts "         ↓           ↓           ↓"
puts "    ┌────────┐  ┌────────┐  ┌────────┐"
puts "    │ Node.js│  │  Ruby  │  │ Rails  │"
puts "    │  API   │  │ Worker │  │  App   │"
puts "    └────────┘  └────────┘  └────────┘"
puts "         ↓           ↓           ↓"
puts "    Same concurrency limiting everywhere!"
puts ""

puts "Problem Solved:"
puts "Before: Ruby + JavaScript = different implementations"
puts "After: One Elide implementation = 100% consistent"
puts ""

puts "Rails Integration (when ready):"
puts "  # config/initializers/p-limit.rb"
puts "  P_LIMIT = Elide.require('./elide-p-limit.ts')"
puts "  "
puts "  # app/services/data_service.rb"
puts "  class DataService"
puts "    def process(data)"
puts "      P_LIMIT.default(data)"
puts "    end"
puts "  end"

File: conversions/p-limit/elide-p-limit.py
#!/usr/bin/env python3
"""
Python Integration Example for elide-p-limit

This demonstrates calling the TypeScript p-limit implementation from Python
using Elide's polyglot capabilities.

Benefits:
- One p-limit implementation shared across TypeScript and Python
- Consistent concurrency limiting behavior across services
- No Python library inconsistencies
- Perfect for data pipelines and processing
"""

# NOTE: Exact syntax depends on Elide's Python polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Python support is ready

# Assuming Elide provides something like:
# from elide import require
# p-limit_module = require('./elide-p-limit.ts')

print("=== Python Calling TypeScript PLimit ===\n")

# Example 1: Basic Usage
# result = p-limit_module.default(input_data)
# print(f"Result: {result}")
# print()

# Example 2: Data Pipeline
# class DataProcessor:
#     def __init__(self):
#         self.p-limit = p-limit_module
#
#     def process_batch(self, items):
#         """Process batch using p-limit"""
#         return [self.p-limit.default(item) for item in items]
#
# processor = DataProcessor()
# data = ['item1', 'item2', 'item3']
# results = processor.process_batch(data)
# for i, result in enumerate(results, 1):
#     print(f"  Item {i}: {result}")
# print()

print("Real-world use case:")
print("- Python data pipeline uses p-limit")
print("- Uses same TypeScript implementation as Node.js API")
print("- Guarantees consistent concurrency limiting across services")
print("- Rate limit API calls")
print()

print("Example: Cross-Service Architecture")
print("┌─────────────────────────────────────────┐")
print("│   Elide PLimit (TypeScript)           │")
print("│   elide-p-limit.ts                        │")
print("└─────────────────────────────────────────┘")
print("         ↓                   ↓")
print("    ┌────────┐          ┌────────┐")
print("    │ Node.js│          │ Python │")
print("    │  API   │          │Pipeline│")
print("    └────────┘          └────────┘")
print("         ↓                   ↓")
print("    Same concurrency limiting everywhere!")
print()

print("Problem Solved:")
print("Before: Python + JavaScript = different concurrency limiting implementations")
print("After: One Elide implementation = 100% consistent behavior")
print()

print("When Elide Python API is ready, usage will be:")
print("  from elide import require")
print("  p-limit = require('./elide-p-limit.ts')")
print("  result = p-limit.default(data)")
